engines/scc_proxy/lib/scc_proxy/engine.rb

```diff
@@ -24,66 +25,107 @@
   Net::HTTPRetriableError
 ].freeze
 
+INSTANCE_ID_KEYS = {
+  Amazon: 'instanceId',
+  Google: 'instance_id',
+  Microsoft: 'vmId'
+}.freeze
+
 # rubocop:disable Metrics/ModuleLength
 module SccProxy
   class << self
 
-    def headers(auth)
+    # rubocop:disable ThreadSafety/ClassAndModuleAttributes
+    attr_accessor :instance_id
+
+    # rubocop:enable ThreadSafety/ClassAndModuleAttributes
+
+    # rubocop:disable ThreadSafety/InstanceVariableInClassMethod
+    def headers(auth, params, logger)
+      @instance_id = if params && params.class != String
+                       get_instance_id(params, logger)
+                     else
+                       # if it is not JSON, it is the system_token already
+                       params
+                     end
+
       {
         'accept' => 'application/json,application/vnd.scc.suse.com.v4+json',
         'Content-Type' => 'application/json',
-        'Authorization' => auth
+        'Authorization' => auth,
+        ApplicationController::SYSTEM_TOKEN_HEADER => @instance_id
       }
     end
+    # rubocop:enable ThreadSafety/InstanceVariableInClassMethod
+
+    def get_instance_id(params, logger)
+      verification_provider = InstanceVerification.provider.new(
+        logger,
+        nil,
+        nil,
+        nil
+      )
+      instance_id_key = INSTANCE_ID_KEYS[params['cloud_provider']]
+      iid = verification_provider.parse_instance_data(params['instance_data'])
```

Comment:
I think you need to explain to me why this whole logic is required here. While I get that there are specific implementations for each provider, I'm wondering what `parse_instance_data` actually does?

Can you pin point me to an implementation wish is used in production?